bundles/org.eclipse.swt/Eclipse SWT PI/win32/library/build.bat

```diff
@@ -51,15 +50,33 @@ IF NOT EXIST "%MSVC_HOME%" (
 @rem Check for a usable JDK
 IF "%SWT_JAVA_HOME%"=="" CALL :ECHO "'SWT_JAVA_HOME' was not provided"
 IF NOT EXIST "%SWT_JAVA_HOME%" (
-    CALL :ECHO "WARNING: x64 Java JDK not found. Please set SWT_JAVA_HOME to the JDK directory containing the intended JDK native headers."
+    CALL :ECHO "WARNING: 64-bit Java JDK not found. Please set SWT_JAVA_HOME to the JDK directory containing the intended JDK native headers."
 )
 
 @rem -----------------------
-set PROCESSOR_ARCHITECTURE=AMD64
-IF "x.%OUTPUT_DIR%"=="x." set OUTPUT_DIR=..\..\..\org.eclipse.swt.win32.win32.x86_64
+IF NOT "%XC_HOST_TARGET%"=="" (
+  @rem Accept the already-set variable value for host_target, for manual cross-compile
+  CALL :ECHO "Manual cross-compile of host_target: %XC_HOST_TARGET%"
+  SET BUILD_ARCH=%XC_HOST_TARGET%
+  @rem With the knowledge that the OUTPUT_DIR is also already set
+  CALL :ECHO "Cross-compile output dir: %OUTPUT_DIR%"
+  IF "x.%OUTPUT_DIR%"=="x." (
+    CALL :ECHO "ERROR: OUTPUT_DIR environment variable must be set for manual cross-compile."
+    EXIT /B 1
+  )
```

Comment:
I think that this code can be simplified.

Realistically I don't expect any new archs in upcoming years, so there will be just two - amd64 and arm64. So I suggest to have just one new variable that would override `%PROCESSOR_ARCHITECTURE%`. For example, `COMPILE_ARCH` could be the only new variable. If it's empty, it falls back to current processor arch:
```
IF "%COMPILE_ARCH%"=="" SET COMPILE_ARCH=%PROCESSOR_ARCHITECTURE%
```
this way, you only use one var instead of two, and there's less code.

Reply:
Doing it this way means the cross-compiling idea is abandoned. Is this what you really want?

The required OUTPUT_DIR variable must be set, either internally within this `build.bat` file, or externally set by the caller of this `build.bat` script. Perhaps we can make this `build.bat` script smarter, in working out the required OUTPUT_DIR value from the COMPILE_ARCH (or XC_HOST_TARGET) value, by parsing either `x64_arm64` or `arm64_x64` and working out the needed OUTPUT_DIR value.

Reply:
Cross-compiled arm64 is the same as native-compiled arm64. Therefore, if arm64 is compiled, output dirs will be for arm64.

Reply:
Nope, on an x64 box, passing 'arm64' to initialize and compile the C code for Arm64 will certainly fail to work (I've tried it). For cross-compile, MSVC must know from which host to which target explicitly, hence it requires 'x64_arm64' on an x64 box, or 'arm64_x64' on an Arm64 box, for cross-compiling.

Reply:
ok, so the expected value for `BUILD_ARCH` is for example `x64_arm64`, I didn't catch that initially. I think that it still makes sense to compose `BUILD_ARCH` from `%PROCESSOR_ARCHITECTURE%` and requested target arch, because the code would be easier to understand. What do you think?

```
REM Compose host architecture string for MSVC
IF "%PROCESSOR_ARCHITECTURE%"=="AMD64" (
  SET HOST_ARCH=x64
) ELSE IF "%PROCESSOR_ARCHITECTURE%"=="ARM64" (
  SET HOST_ARCH=arm64
) ELSE (
  CALL :ECHO "ERROR: Unknown host architecture: %PROCESSOR_ARCHITECTURE%."
  EXIT /B 1
)

REM %TARGET_ARCH% may be specified by the caller for cross-compiling.
REM If not, build for builder machine's architecture
IF "%TARGET_ARCH%"=="" (
  SET  TARGET_ARCH=%HOST_ARCH%
)

REM Compose build argument for MSVC
IF "%TARGET_ARCH%"=="%HOST_ARCH%" (
  SET BUILD_ARCH=%TARGET_ARCH%
) ELSE (
   SET BUILD_ARCH=%HOST_ARCH%_%TARGET_ARCH%
)

REM Select build directory based on target arch
IF "%TARGET_ARCH%"=="x64" (
  IF "x.%OUTPUT_DIR%"=="x." SET OUTPUT_DIR=..\..\..\org.eclipse.swt.win32.win32.x86_64
) ELSE IF "%TARGET_ARCH%"=="arm64" (
  IF "x.%OUTPUT_DIR%"=="x." SET OUTPUT_DIR=..\..\..\org.eclipse.swt.win32.win32.aarch64
) ELSE (
  CALL :ECHO "ERROR: Unknown target architecture: %TARGET_ARCH%."
  EXIT /B 1
)
```

Reply:
Thinking about it some more, I figured that my approach has an even bigger advantage: user only specifies what he wants, and doesn't have to care about how to make this happen. Script will figure that it's cross-compiling on its own.

Reply:
This, in turn, will simplify builder scripts: that will just say "build that and that", regardless of what they're running on.

Reply:
OK, your approach is a bit verbose, but it's quite clear in intention, so I have no problem with it. And current Maven build scripts would still work with it without change (this was my intention all along.) And doing cross-compiling has to be a _manual_ operation anyway, that the user need to explicitly specify an extra environment variable in order to trigger it, so I won't expect the current automated build scripts to do this cross-compiling, but the facility is there for automated cross-compile by the build scripts if required.

I'll merge this with the PR and squash it. And the PR's description need be changed too, to reflect the updated contents.

Another thing: I plan to apply these changes of `build.bat`, to the [build.bat](https://github.com/eclipse-equinox/equinox/blob/master/features/org.eclipse.equinox.executable.feature/library/win32/build.bat) of the Equinox project, to generate the Equinox launcher native for Arm64. If you have different idea (for the Equinox project), let me know in advance, thanks.